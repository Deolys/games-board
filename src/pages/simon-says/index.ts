export { SimonSaysGame } from "./simon-says";