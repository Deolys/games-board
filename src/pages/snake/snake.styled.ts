import styled from "@emotion/styled";
import { Box } from "@mui/material";

export const GameArea=styled(Box)({
  width: '400px',
  height: '400px',
  margin: '20px auto',
  position: 'relative',
  border: '2px solid #000',
});

export const SnakeDot=styled(Box)({
  width: '20px',
  height: '20px',
  backgroundColor: 'green',
  position: 'absolute',
});