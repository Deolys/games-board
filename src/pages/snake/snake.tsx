import { useState, useEffect, useCallback } from 'react';
import { AppBar, Toolbar, Typography, Paper, Snackbar, Alert } from '@mui/material';
import { Header } from '../../components/header';
import { GameArea, SnakeDot } from './snake.styled';

type Direction = 'RIGHT' | 'LEFT' | 'UP' | 'DOWN';

export function SnakeGame() {
  const [snakeDots, setSnakeDots] = useState<number[][]>([[10, 10]]);
  const [direction, setDirection] = useState<Direction>('RIGHT');
  const [food, setFood] = useState<number[]>([4, 4]);
  const speed = 200;
  const [gameOver, setGameOver] = useState(false);
  const [open, setOpen] = useState(false);
  const [message, setMessage] = useState('');
  
    const onGameOver = (): void => {
      setGameOver(true);
    };

    useEffect(() => {  
      if (gameOver) {
        setOpen(true);
        setMessage(`Игра окончена. Ваш счет: ${snakeDots.length}`);
        setSnakeDots([[10,10]]);
        setDirection('RIGHT');
        setFood([4,4]);
        setGameOver(false);
      }
    }, [gameOver, snakeDots.length])

    const checkIfEat = useCallback((head: number[], foodPosition: number[]): boolean => {

      if (head[0] === foodPosition[0] && head[1] === foodPosition[1]) {
        let newFoodPosition: number[];
        do {
          newFoodPosition = [Math.floor(Math.random() * (20)), Math.floor(Math.random() * (20))];
        } while (snakeDots.some(dot => dot[0] === newFoodPosition[0] && dot[1] === newFoodPosition[1]));
        setFood(newFoodPosition);
        return true;
      }
      return false;
    }, [snakeDots]);
  
  useEffect(() => {
    const onKeyDown = (e: KeyboardEvent): void  => {
      switch (e.key) {
        case 'ArrowUp':
          setDirection('UP');
          break;
        case 'ArrowDown':
          setDirection('DOWN');
          break;
        case 'ArrowLeft':
          setDirection('LEFT');
          break;
        case 'ArrowRight':
          setDirection('RIGHT');
          break;
      }
    };

    document.addEventListener('keydown', onKeyDown);

    return () => {
      document.removeEventListener('keydown', onKeyDown);
    };
  }, []);

  const moveSnake = useCallback(() => {
    const newSnakeDots = [...snakeDots];
    let head = newSnakeDots[newSnakeDots.length - 1];

    switch (direction) {
      case 'RIGHT':
        head = [head[0] + 1, head[1]];
        break;
      case 'LEFT':
        head = [head[0] - 1, head[1]];
        break;
      case 'DOWN':
        head = [head[0], head[1] + 1];
        break;
      case 'UP':
        head = [head[0], head[1] - 1];
        break;
      default:
        return;
    }

    const snakeBody = newSnakeDots.slice(0, -1);
    snakeBody.forEach((dot) => {
        if (head[0] === dot[0] && head[1] === dot[1]) {
          onGameOver();
        }
      });

    if (head[0] >= 20 || head[1] >= 20 || head[0] < 0 || head[1] < 0) {
      onGameOver();
    }
    
    newSnakeDots.push(head);
    newSnakeDots.shift();
    setSnakeDots(newSnakeDots);
    if (checkIfEat(head, food))
      setSnakeDots(prevSnake => [[], ...prevSnake]);

  }, [direction, food, snakeDots, checkIfEat]);

  useEffect(() => {
    const interval = setInterval(moveSnake, speed);
    return () => clearInterval(interval);
  }, [moveSnake]);

  const handleClose = (_event?: React.SyntheticEvent | Event, reason?: string) => {
    if (reason === 'clickaway') {
      return;
    }
    setOpen(false);
  };

  let touchStartX: number | null = null;
  let touchStartY: number | null = null;

const handleTouchStart = (e: React.TouchEvent): void  => {
  touchStartX = e.changedTouches[0].screenX;
  touchStartY = e.changedTouches[0].screenY;
};

const handleTouchMove = (e: React.TouchEvent): void  => {
  if (!touchStartX || !touchStartY) {
    return;
  }

  const touchEndX = e.changedTouches[0].screenX;
  const touchEndY = e.changedTouches[0].screenY;

  const diffX = touchStartX - touchEndX;
  const diffY = touchStartY - touchEndY;

  if (Math.abs(diffX) > Math.abs(diffY)) {
    if (diffX > 0) {
      setDirection('LEFT');
    } else {
      setDirection('RIGHT');
    }
  } else {
    if (diffY > 0) {
      setDirection('UP');
    } else {
      setDirection('DOWN');
    }
  }

  e.preventDefault();
};
  
    return (
      <>
      <Header />
      <Paper onTouchStart={handleTouchStart} onTouchMove={handleTouchMove} data-testid='game-zone'>
        <AppBar position="static">
          <Toolbar>
            <Typography variant="h6">
              Игра "Змейка" на React с Material-UI
            </Typography>
          </Toolbar>
        </AppBar>
        <Snackbar open={open} autoHideDuration={4000} onClose={handleClose} anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}>
          <Alert onClose={handleClose} severity="info" sx={{ width: '100%' }}>
            {message}
          </Alert>
        </Snackbar>
        <GameArea>
          {snakeDots.map((dot, index) => {
            const style = {
              left: `${dot[0]*20}px`,
              top: `${dot[1]*20}px`
            };

            return <SnakeDot key={index} style={style} data-testid="snake-dot"/>;
          })}
          <SnakeDot style={{ left: `${food[0]*20}px`, top: `${food[1]*20}px`, backgroundColor: 'red' }} />
        </GameArea>
      </Paper>
      </>
    );
  }
  
  export default SnakeGame;