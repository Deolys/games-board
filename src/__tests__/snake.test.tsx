import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { SnakeGame } from "../pages/snake";
import '@testing-library/jest-dom';

describe('SnakeGame', () => {
  it('SnakeGame mounts properly', () => {
    const { container } = render(<SnakeGame />);
    expect(container).toMatchSnapshot();
  })

  beforeEach(() => {
    render(<SnakeGame />);
  })

  it('The snake moves up, when the ArrowUp key is pressed', async () => {
    const snakeDot = screen.getByTestId('snake-dot');

    // Симулируем нажатие клавиши "ArrowUp"
    fireEvent.keyDown(document, { key: 'ArrowUp', code: 'ArrowUp' });

    // Проверяем, что голова змеи переместилась вверх
     await waitFor(() => {
      expect(snakeDot.style.top).toBe('180px');
    });
  });

  it('The snake moves down, when the ArrowDown key is pressed', async () => {
    const snakeDot = screen.getByTestId('snake-dot');

    // Симулируем нажатие клавиши "ArrowDown"
    fireEvent.keyDown(document, { key: 'ArrowDown', code: 'ArrowDown' });

    // Проверяем, что голова змеи переместилась вниз
     await waitFor(() => {
      expect(snakeDot.style.top).toBe('220px');
    });
  });

  it('The snake moves left, when the ArrowLeft key is pressed', async () => {
    const snakeDot = screen.getByTestId('snake-dot');

    // Симулируем нажатие клавиши "ArrowLeft"
    fireEvent.keyDown(document, { key: 'ArrowLeft', code: 'ArrowLeft' });

    // Проверяем, что голова змеи переместилась влево
     await waitFor(() => {
      expect(snakeDot.style.left).toBe('180px');
    });
  });

  it('The snake moves right, when the ArrowRight key is pressed', async () => {
    const snakeDot = screen.getByTestId('snake-dot');

    // Симулируем нажатие клавиши "ArrowRight"
    fireEvent.keyDown(document, { key: 'ArrowRight', code: 'ArrowRight' });

    // Проверяем, что голова змеи переместилась вправо
     await waitFor(() => {
      expect(snakeDot.style.left).toBe('220px');
    });
  });

  it('The snake moves up correctly on mobile devices', async () => {
    const snakeDot = screen.getByTestId('snake-dot');
    const gameZone = screen.getByTestId('game-zone');

    fireEvent.touchStart(gameZone, {
      changedTouches: [{ screenX: 50, screenY: 100 }]
    });

    fireEvent.touchMove(gameZone, {
      changedTouches: [{ screenX: 50, screenY: 50 }]
    });

    // Проверяем, что голова змеи переместилась вверх
     await waitFor(() => {
      expect(snakeDot.style.top).toBe('180px');
    });
  })
})