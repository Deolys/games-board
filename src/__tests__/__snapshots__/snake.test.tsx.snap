// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnakeGame SnakeGame mounts properly 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1413kmy-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-u4fbkj-MuiToolbar-root"
    >
      <a
        class="css-3qnz2e"
        href="/"
      >
        <img
          alt="Games board"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-1waifce-MuiTypography-root"
        >
          Games board
        </p>
      </a>
      <div
        class="MuiStack-root css-1dzrg60-MuiStack-root"
      >
        <a
          class="MuiTypography-root MuiTypography-body1 css-1v5m6us-MuiTypography-root"
          href="/"
        >
          HOME
        </a>
        <button
          aria-haspopup="true"
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-17ovnxk-MuiButtonBase-root-MuiButton-root"
          id="composition-button"
          tabindex="0"
          type="button"
        >
          Games
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </header>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
    data-testid="game-zone"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
        >
          Игра "Змейка" на React с Material-UI
        </h6>
      </div>
    </header>
    <div
      class="MuiBox-root css-plrrfd"
    >
      <div
        class="MuiBox-root css-oim5cv"
        data-testid="snake-dot"
        style="left: 200px; top: 200px;"
      />
      <div
        class="MuiBox-root css-oim5cv"
        style="left: 80px; top: 80px; background-color: red;"
      />
    </div>
  </div>
</div>
`;
